File: src/AutoContext.Framework.Web/src/transport/pipe-keep-alive-client.ts
import type { Socket } from 'node:net';

import type { Logger } from '../logging/logger.js';
import { PipeTransport } from './pipe-transport.js';

/**
 * Layer-3 keep-alive client: connects, writes a tiny handshake once,
 * and holds the socket open as a process liveness signal. The remote
 * peer treats the live socket as "this host is running" and observes
 * the close when the host exits.
 *
 * When the pipe name is empty (standalone runs, no parent listening)
 * {@link start} is a no-op so call sites don't need to special-case
 * standalone scenarios. Connection failures are best-effort — they're
 * logged at warning and swallowed.
 *
 * Counterpart of the C# `PipeKeepAliveClient` in
 * `AutoContext.Framework`.
 */
export class PipeKeepAliveClient {
    private static readonly DEFAULT_CONNECT_TIMEOUT_MS = 2000;

    private readonly transport: PipeTransport;
    private readonly logger: Logger;
    private socket: Socket | undefined;
    private startPromise: Promise<void> | undefined;
    private disposed = false;

    constructor(transport: PipeTransport, logger: Logger) {
        this.transport = transport;
        this.logger = logger;
    }

    /**
     * Connects to {@link pipeName} and writes {@link handshake} as a
     * single best-effort attempt. On success the socket is held until
     * {@link dispose}. Safe to call multiple times — concurrent callers
     * share a single completion promise.
     */
    start(
        pipeName: string,
        handshake: Buffer | Uint8Array,
        connectTimeoutMs: number = PipeKeepAliveClient.DEFAULT_CONNECT_TIMEOUT_MS,
    ): Promise<void> {
        if (this.startPromise !== undefined) {
            return this.startPromise;
        }
        this.startPromise = this.startCore(pipeName, handshake, connectTimeoutMs);
        return this.startPromise;
    }

    /**
     * Closes the held socket so the remote peer observes the
     * disconnect. Safe to call multiple times.
     */
    async dispose(): Promise<void> {
        if (this.disposed) {
            return;
        }
        this.disposed = true;

        const socket = this.socket;
        this.socket = undefined;
        if (socket !== undefined) {
            await new Promise<void>((resolve) => {
                socket.once('close', () => resolve());
                socket.destroy();
            });
        }
    }

    private async startCore(
        pipeName: string,
        handshake: Buffer | Uint8Array,
        connectTimeoutMs: number,
    ): Promise<void> {
        if (pipeName === '') {
            this.logger.debug('Keep-alive pipe not configured; skipping connection.');
            return;
        }

        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), connectTimeoutMs);
        timer.unref();

        try {
            const socket = await this.transport.connect(pipeName, controller.signal);

            if (this.disposed) {
                socket.destroy();
                return;
            }

            if (handshake.length > 0) {
                await new Promise<void>((resolve, reject) => {
                    socket.write(handshake, (err) => {
                        if (err !== null && err !== undefined) {
                            reject(err);
                            return;
                        }
                        resolve();
                    });
                });
            }

            // Swallow late socket errors — keep-alive is best-effort.
            socket.on('error', () => { /* see class doc */ });

            this.socket = socket;
        }
        catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            this.logger.warn(`Failed to connect keep-alive pipe '${pipeName}': ${message}`);
        }
        finally {
            clearTimeout(timer);
        }
    }
}

File: src/AutoContext.Framework/Transport/PipeKeepAliveClient.cs
namespace AutoContext.Framework.Transport;

using System.IO.Pipes;

using Microsoft.Extensions.Logging;

/// <summary>
/// Layer-3 keep-alive client: connects, writes a tiny handshake once,
/// and holds the socket open as a process liveness signal. The remote
/// peer treats the live socket as "this host is running" and observes
/// the close when the host exits.
/// </summary>
/// <remarks>
/// When the pipe name is empty (standalone runs, no parent listening)
/// <see cref="StartAsync"/> is a no-op so call sites don't need to
/// special-case standalone scenarios. Connection failures are
/// best-effort — they're logged at <c>Warning</c> and swallowed.
/// </remarks>
public sealed partial class PipeKeepAliveClient : IAsyncDisposable
{
    private readonly PipeTransport _transport;
    private readonly ILogger<PipeKeepAliveClient> _logger;
    private Stream? _stream;
    private bool _disposed;

    /// <summary>
    /// Creates a new <see cref="PipeKeepAliveClient"/>. Both arguments
    /// are required.
    /// </summary>
    public PipeKeepAliveClient(PipeTransport transport, ILogger<PipeKeepAliveClient> logger)
    {
        ArgumentNullException.ThrowIfNull(transport);
        ArgumentNullException.ThrowIfNull(logger);

        _transport = transport;
        _logger = logger;
    }

    /// <summary>
    /// Connects to <paramref name="pipeName"/> and writes
    /// <paramref name="handshake"/> as a single best-effort attempt.
    /// On success the socket is held until <see cref="DisposeAsync"/>;
    /// on failure the warning is logged and the call returns normally.
    /// </summary>
    public async Task StartAsync(
        string pipeName,
        ReadOnlyMemory<byte> handshake,
        int connectTimeoutMs = 2000,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(pipeName);

        if (_disposed)
        {
            return;
        }

        if (pipeName.Length == 0)
        {
            LogSkippingNoPipe(_logger);
            return;
        }

        Stream? stream = null;
        try
        {
            stream = await _transport.ConnectAsync(
                pipeName, connectTimeoutMs, PipeDirection.Out, cancellationToken).ConfigureAwait(false);

            if (handshake.Length > 0)
            {
                await stream.WriteAsync(handshake, cancellationToken).ConfigureAwait(false);
                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
            }

            if (_disposed)
            {
                await stream.DisposeAsync().ConfigureAwait(false);
                return;
            }

            _stream = stream;
            stream = null;
        }
        catch (OperationCanceledException)
        {
            if (stream is not null)
            {
                await stream.DisposeAsync().ConfigureAwait(false);
                stream = null;
            }
            throw;
        }
        catch (Exception ex) when (ex is TimeoutException or IOException or UnauthorizedAccessException)
        {
            LogConnectFailed(_logger, pipeName, ex);
        }
        finally
        {
            if (stream is not null)
            {
                await stream.DisposeAsync().ConfigureAwait(false);
            }
        }
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;

        var stream = _stream;
        _stream = null;
        if (stream is not null)
        {
            await stream.DisposeAsync().ConfigureAwait(false);
        }
    }

    [LoggerMessage(EventId = 1, Level = LogLevel.Debug,
        Message = "Keep-alive pipe not configured; skipping connection.")]
    private static partial void LogSkippingNoPipe(ILogger logger);

    [LoggerMessage(EventId = 2, Level = LogLevel.Warning,
        Message = "Failed to connect keep-alive pipe '{PipeName}'.")]
    private static partial void LogConnectFailed(ILogger logger, string pipeName, Exception ex);
}
